File: 백준/Silver/18404. 현명한 나이트/현명한 나이트.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    // N and M
    var a = br.readLine().split(" ").map { it.toInt() }
    val N = a[0]
    val M = a[1]

    // X and Y
    a = br.readLine().split(" ").map { it.toInt() }
    val X = a[0]-1
    val Y = a[1]-1

    // bfs
    val dy = arrayOf(-1, 1, -2, 2, -2, 2, -1, 1)
    val dx = arrayOf(-2, -2, -1, -1, 1, 1, 2, 2)

    val visited = Array<IntArray>(N) {IntArray(N) }

    val queue = LinkedList<Pair<Int,Int>>()
    visited[Y][X] = 1
    queue.addLast(Pair(Y,X))    // offer
    while(queue.isNotEmpty()) {
        val now = queue.pollFirst()

        for (i in 0 until 8) {
            val ny = now.first + dy[i]
            val nx = now.second + dx[i]
            if(ny<0 || ny>=N || nx<0 || nx>=N) continue
            if(visited[ny][nx] > 0) continue
            visited[ny][nx] = visited[now.first][now.second] + 1
            queue.addLast(Pair(ny,nx))  // offer
        }
    }

    // Question
    (0 until M).forEach {
        a = br.readLine().split(" ").map { it.toInt() }
        bw.write("${visited[a[1]-1][a[0]-1]-1} ")
    }

    br.close();
    bw.flush();
    bw.close();
}

File: 백준/Silver/18404. 현명한 나이트/현명한 나이트.cc
#include <bits/stdc++.h>
using namespace std;

int N, M, X, Y, visited[504][504];
int dy[] = {-1, 1, -2, 2, -2, 2, -1, 1};
int dx[] = {-2, -2, -1, -1, 1, 1, 2, 2};

int main(void) {
    ios_base::sync_with_stdio(0); cin.tie(0); cout.tie(0);

    cin >> N >> M;
    cin >> X >> Y;

    visited[Y-1][X-1] = 1;
    queue<pair<int,int>> q;
    q.push({Y-1,X-1});
    while(q.size()) {
        auto it = q.front(); q.pop();
        
        for(int i = 0; i < 8; i++) {
            int ny = it.first + dy[i];
            int nx = it.second + dx[i];
            if(ny<0 || ny>=N || nx<0 || nx>=N) continue;
            if(visited[ny][nx]) continue;
            visited[ny][nx] = visited[it.first][it.second] + 1;
            q.push({ny,nx});
        }
    }

    while(M--) {
        cin >> X >> Y;
        cout << visited[Y-1][X-1]-1 << ' ';
    }
}